posts/inside-rust/2020-05-25-new-inline-asm.md

```diff
@@ -0,0 +1,130 @@
+  internals](https://internals.rust-lang.org/t/pre-rfc-2-inline-assembly/11310).
+- Inline assembly became the language team's first [project
+  group](https://github.com/rust-lang/rfcs/blob/master/text/2836-project-asm.md),
+  and iteratively designed RFCs in [the project group
+  repository](https://github.com/rust-lang/project-inline-asm/).
+- [RFC 2873](https://github.com/rust-lang/rfcs/pull/2873) (still under
+  discussion) provides a specification for the syntax and its interaction with
+  the Rust language.
+- We [renamed the existing `asm!` to
+  `llvm_asm!`](https://github.com/rust-lang/rust/pull/68404), so that people
+  currently using inline assembly on nightly can continue to use the existing
+  syntax for now. (We plan to remove this syntax eventually, given its fragile
+  ICE-happy nature, but while evaluating the new syntax we want the old syntax
+  available for comparison and alternatives.)
+- [PR 69171](https://github.com/rust-lang/rust/pull/69171) (also by Amanieu)
+  implemented the new `asm!` syntax in nightly.
+
+Here's an example of using the new inline assembly syntax, to print a message
+to standard output using a direct [`write`
+syscall](https://man7.org/linux/man-pages/man2/write.2.html) on x86-64 Linux:
+
+```rust
+#![feature(asm)]
+
+fn main() {
+    let buf = "Hello from asm!\n";
+    let ret: i32;
+    unsafe {
+        asm!(
+            "syscall",
+            in("rax") 1, // syscall number
+            in("rdi") 1, // fd
+            in("rsi") buf.as_ptr(),
+            in("rdx") buf.len(),
+            out("rcx") _, // clobbered by syscalls
+            out("r11") _, // clobbered by syscalls
+            lateout("rax") ret,
+        );
+    }
+    println!("write returned: {}", ret);
+}
+```
+
+(You can [try this example on the
+playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=e983a5f5cffa51f4320f1176465d3a56).)
+
+The example above specifies the exact inputs, outputs, and clobbers required by
+the Linux syscall calling convention. You can also provide inputs and outputs
+via arbitrary registers, and the compiler will select appropriate registers for
+you. The following example uses [bit manipulation
+instructions](https://en.wikipedia.org/wiki/Bit_Manipulation_Instruction_Sets)
+to compute the bit numbers of all set bits in a value, and stores them in a
+slice of memory:
+
+```rust
+#![feature(asm)]
+
+fn main() {
+    let mut bits = [0u8; 64];
+    for value in 0..=1024u64 {
+        let popcnt;
+        unsafe {
+            asm!("
+                popcnt {popcnt}, {v}
+                2:
+                blsi rax, {v}
+                jz 1f
+                xor {v}, rax
+                tzcnt rax, rax
+                stosb
+                jmp 2b
+                1:
+                ",
+                v = inout(reg) value => _,
+                popcnt = out(reg) popcnt,
+                out("rax") _, // scratch
+                inout("rdi") bits.as_mut_ptr() => _,
```

Comment:
`options(nostack)` could be used here.

Reply:
This doesn't seem to make any difference in the generated code.

Along the same lines, I could use `option(readonly)` in the `syscall` example, but I don't think doing so adds any value. These are supposed to be very simple examples, and people will have the full RFC available for reference.